common/NMEA_parser.h

```diff
@@ -0,0 +1,131 @@
+// NMEA message parser
+
+/*
+          1       2      3 4        5 6 7  8   9  10 11  12 13 14   15
+|         |       |      | |        | | |  |   |   | |   | |   |    |
+$--GGA,hhmmss.ss,llll.ll,a,yyyyy.yy,a,x,xx,x.x,x.x,M,x.x,M,x.x,xxxx*hh
+
+1) Time (UTC)
+2) Latitude
+3) N or S (North or South)
+4) Longitude
+5) E or W (East or West)
+6) GPS Quality Indicator,
+0 = Invalid, 1 = Valid SPS, 2 = Valid DGPS, 3 = Valid PPS
+7) Number of satellites in view, 00 - 12
+8) Horizontal Dilution of precision, lower is better
+9) Antenna Altitude above/below mean-sea-level (geoid)
+10) Units of antenna altitude, meters
+11) Geoidal separation, the difference between the WGS-84 earth
+ellipsoid and mean-sea-level (geoid), '-' means mean-sea-level below ellipsoid
+12) Units of geoidal separation, meters
+13) Age of differential GPS data, time in seconds since last SC104
+type 1 or 9 update, null field when DGPS is not used
+14) Differential reference station ID, 0000-1023
+15) Checksum
+*/
+#pragma once
+#include<string>
+#include<vector>
+#include "../third_party/units.h"
+
+class NMEA_parser {
+    
+    using meter_t = units::length::meter_t;
+    using degree_t = units::angle::degree_t;
+    using arcminute_t = units::angle::arcminute_t;
+    using meters_per_second_t = units::velocity::meters_per_second_t;
+    using knot_t = units::velocity::knot_t;
+    
+
+    private:
+
+    static std::vector<std::string> parseNMEAstring(const std::string textToParse, const std::string NMEA_sentence_id) {
+        using namespace std;
+        const string delimiter = "$";      // sentences separated by [$]
+        const string w_delimiter = ",";    // elements separated by  [,]
+        string s = textToParse;
+
+        size_t pos = 0;
+        string token;
+        vector<string> tokens,words;
+
+        // separating the lines 
+        while ((pos = s.find(delimiter)) != string::npos) {
+            token = s.substr(0, pos);
+            tokens.push_back(token);
+            s.erase(0, pos + delimiter.length());
+        }
+
+        // separating the words
+        for (auto &f : tokens) {         
+            string word;
+            pos = 0;   
+            while ((pos = f.find(w_delimiter)) != string::npos) {
+                word = f.substr(0,pos);
+                words.push_back(word);
+                f.erase(0, pos + w_delimiter.length());         
+            }
+
+            // if we find the one, we stop 
+            if (!words.empty() && words.at(0) == NMEA_sentence_id) {          
```

Comment:
Use ``words[0]`` instead. The ``at()`` method performs a bounds check and throws an exception if it's out of range, but we know there will be at least one element because you're checking for it.